heat/core/tests/test_manipulations.py

```diff
@@ -315,6 +316,273 @@ def test_concatenate(self):
         with self.assertRaises(RuntimeError):
             ht.concatenate((ht.zeros((2, 2), split=0), ht.zeros((2, 2), split=1)), axis=0)
 
+    def test_diag(self):
+        size = ht.MPI_WORLD.size
+        rank = ht.MPI_WORLD.rank
+
+        data = torch.arange(size * 2)
+        a = ht.array(data)
+        res = ht.diag(a)
+        self.assertTrue(torch.equal(res._DNDarray__array, torch.diag(data)))
+
+        res = ht.diag(a, offset=size)
+        self.assertTrue(torch.equal(res._DNDarray__array, torch.diag(data, diagonal=size)))
+
+        res = ht.diag(a, offset=-size)
+        self.assertTrue(torch.equal(res._DNDarray__array, torch.diag(data, diagonal=-size)))
+
+        a = ht.array(data, split=0)
+        res = ht.diag(a)
+        self.assertEqual(res.split, a.split)
+        self.assertEqual(res.shape, (size * 2, size * 2))
+        self.assertEqual(res.lshape[res.split], 2)
+        exp = torch.diag(data)
+        for i in range(rank * 2, (rank + 1) * 2):
+            self.assertTrue(torch.equal(res[i, i]._DNDarray__array, exp[i, i]))
+
+        res = ht.diag(a, offset=size)
+        self.assertEqual(res.split, a.split)
+        self.assertEqual(res.shape, (size * 3, size * 3))
+        self.assertEqual(res.lshape[res.split], 3)
+        exp = torch.diag(data, diagonal=size)
+        for i in range(rank * 3, min((rank + 1) * 3, a.shape[0])):
+            self.assertTrue(torch.equal(res[i, i + size]._DNDarray__array, exp[i, i + size]))
+
+        res = ht.diag(a, offset=-size)
+        self.assertEqual(res.split, a.split)
+        self.assertEqual(res.shape, (size * 3, size * 3))
+        self.assertEqual(res.lshape[res.split], 3)
+        exp = torch.diag(data, diagonal=-size)
+        for i in range(max(size, rank * 3), (rank + 1) * 3):
+            self.assertTrue(torch.equal(res[i, i - size]._DNDarray__array, exp[i, i - size]))
+
+        self.assertTrue(ht.equal(ht.diag(ht.diag(a)), a))
+
+        a = ht.random.rand(15, 20, 5, split=1)
+        res_1 = ht.diag(a)
+        res_2 = ht.diagonal(a)
+        self.assertTrue(ht.equal(res_1, res_2))
+
+        with self.assertRaises(ValueError):
+            ht.diag(data)
+
+        with self.assertRaises(ValueError):
+            ht.diag(a, offset=None)
+
+        a = ht.arange(size)
+        with self.assertRaises(ValueError):
+            ht.diag(a, offset="3")
+
+        a = ht.empty([])
+        with self.assertRaises(ValueError):
+            ht.diag(a)
+
+        if rank == 0:
+            data = torch.ones(size, dtype=torch.int32)
+        else:
+            data = torch.empty(0, dtype=torch.int32)
+        a = ht.array(data, is_split=0)
+        res = ht.diag(a)
+        self.assertTrue(
+            torch.equal(res[rank, rank]._DNDarray__array, torch.tensor(1, dtype=torch.int32))
+        )
```

Comment:
please add a dtype test here to ensure that the dtype is not changed by the function, and that it handles different dtypes